langserver/fs.go

```diff
@@ -26,42 +26,62 @@ func IsFileSystemRequest(method string) bool {
 		method == "textDocument/didSave"
 }
 
-func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) error {
+// HandleFileSystemRequest handles textDocument/did* requests. The path the
+// request is for is returned. true is returned if a file was modified.
+func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) (string, bool, error) {
 	span := opentracing.SpanFromContext(ctx)
 	h.Mu.Lock()
 	overlay := h.overlay
 	h.Mu.Unlock()
 
+	do := func(uri string, op func() error) (string, bool, error) {
+		span.SetTag("uri", uri)
+		before, beforeErr := h.readFile(ctx, uri)
+		err := op()
+		after, afterErr := h.readFile(ctx, uri)
+		if os.IsNotExist(beforeErr) && os.IsNotExist(afterErr) {
+			// Other error conditions we are conservative and
+			// assume something changed.
```

Comment:
This comment could just be `// File didn't exist before or after so nothing has changed`

Then in the else block you can say

```
// If an error prevented us from reading the file before or after
// then we assume the file changed to be conservative.
```